File: Golang/606. Construct String from Binary Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func tree2str(t *TreeNode) string { // 24ms
    if t == nil {
        return ""
    }
    sstring := add2str(t)
    sstrings := strings.Split(sstring, "()")
    sstring = strings.Join(sstrings, "")
    sstring = strings.Replace(sstring, "(left)", "()", -1)
    return sstring[1: len(sstring)-1]
}

func add2str(node *TreeNode) string {
    if node == nil {
        return "()"
    } else if node.Left == nil && node.Right == nil {
        return "(" + strconv.Itoa(node.Val) + ")"
    } else if node.Left != nil {
        return "(" + strconv.Itoa(node.Val) + add2str(node.Left) + add2str(node.Right) + ")"
    } else {
        return "(" + strconv.Itoa(node.Val) + "(left)" + add2str(node.Right) + ")"
    }
}

File: Python3/606. Construct String from Binary Tree.py
class Solution:
    def tree2str(self, t): # 76ms
        """
        :type t: TreeNode
        :rtype: str
        """
        def add2str(node):
            if not node:
                return '()'
            elif not node.left and not node.right:
                return '(' + str(node.val) + ')'
            elif node.left:
                return '(' + str(node.val) + add2str(node.left) + add2str(node.right) + ')'
            else:
                return '(' + str(node.val) + '(left)' + add2str(node.right) + ')'
        
        t = add2str(t)
        t = t.split('()')
        t = ''.join(t)
        t = t.replace('(left)', '()')
        return t[1: -1]
